{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAGNet - PVCA \n",
    "\n",
    "R4.0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(stringr)\n",
    "library(\"SummarizedExperiment\")\n",
    "\n",
    "options(repr.plot.width=16, repr.plot.height=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download data to preferred directory\n",
    "\n",
    "```\n",
    "#!/bin/sh\n",
    "wget --output-document data.zip 'https://data.dieterichlab.org/s/3gCTLGT4DaAqaqb/download' --header 'Accept: text/html' --header 'Connection: keep-alive'\n",
    "unzip -o data.zip\n",
    "rm data.zip\n",
    "rm MAGNetApp/magnetique.sqlite MAGNetApp/data/colData.txt\n",
    "rm -r MAGNetApp/data/DGE MAGNetApp/data/DTU MAGNetApp/data/networks MAGNetApp/data/RBP\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dirloc <- '/prj/MAGE/analysis/genetonic/PVCA'\n",
    "\n",
    "load(file.path(dirloc, 'MAGNetApp', 'data', 'MAGNet.RData'))\n",
    "\n",
    "colData$AFib <- factor(colData$AFib, levels=c(\"No\", \"Yes\"))\n",
    "colData$VTVF <- factor(colData$VTVF, levels=c(\"No\", \"Yes\"))\n",
    "colData$Diabetes <- factor(colData$Diabetes, levels=c(\"No\", \"Yes\"))\n",
    "colData$Hypertension <- factor(colData$VTVF, levels=c(\"No\", \"Yes\"))\n",
    "colData$LibraryPool <- factor(colData$LibraryPool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 20</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Run</th><th scope=col>Experiment</th><th scope=col>LibraryPool</th><th scope=col>TIN</th><th scope=col>RIN</th><th scope=col>DuplicationRate</th><th scope=col>TissueSource</th><th scope=col>Etiology</th><th scope=col>Race</th><th scope=col>Age</th><th scope=col>Sex</th><th scope=col>Weight</th><th scope=col>Height</th><th scope=col>HW</th><th scope=col>LVMass</th><th scope=col>AFib</th><th scope=col>VTVF</th><th scope=col>Diabetes</th><th scope=col>Hypertension</th><th scope=col>LVEF</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>C00039</th><td>SRR10676821</td><td>SRX7354049</td><td>Magnet_10</td><td>72.76478</td><td>8.4</td><td>0.321948</td><td>NF</td><td>NFD</td><td>AA</td><td>18</td><td>M</td><td>70</td><td>175</td><td> NA</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>0.37</td></tr>\n",
       "\t<tr><th scope=row>C00055</th><td>SRR10676822</td><td>SRX7354050</td><td>Magnet_11</td><td>74.29391</td><td>9.1</td><td>0.461090</td><td>NF</td><td>NFD</td><td>AA</td><td>26</td><td>M</td><td>85</td><td>183</td><td> NA</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>  NA</td></tr>\n",
       "\t<tr><th scope=row>C00074</th><td>SRR10676823</td><td>SRX7354051</td><td>Magnet_09</td><td>77.44517</td><td>7.8</td><td>0.338003</td><td>NF</td><td>NFD</td><td>AA</td><td>17</td><td>M</td><td>68</td><td>173</td><td>400</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>0.27</td></tr>\n",
       "\t<tr><th scope=row>C00085</th><td>SRR10676824</td><td>SRX7354319</td><td>Magnet_06</td><td>69.51659</td><td>9.4</td><td>0.420441</td><td>NF</td><td>NFD</td><td>AA</td><td>59</td><td>F</td><td>66</td><td>157</td><td>380</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>0.60</td></tr>\n",
       "\t<tr><th scope=row>C00105</th><td>SRR10676825</td><td>SRX7354320</td><td>Magnet_03</td><td>73.60743</td><td>8.8</td><td>0.290037</td><td>NF</td><td>NFD</td><td>AA</td><td>59</td><td>F</td><td>87</td><td>157</td><td> NA</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>0.55</td></tr>\n",
       "\t<tr><th scope=row>C00120</th><td>SRR10676826</td><td>SRX7354321</td><td>Magnet_12</td><td>30.35077</td><td>8.6</td><td>0.595905</td><td>NF</td><td>NFD</td><td>C </td><td>50</td><td>M</td><td>87</td><td>175</td><td>640</td><td>NA</td><td>No</td><td>No</td><td>No</td><td>No</td><td>0.65</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 x 20\n",
       "\\begin{tabular}{r|llllllllllllllllllll}\n",
       "  & Run & Experiment & LibraryPool & TIN & RIN & DuplicationRate & TissueSource & Etiology & Race & Age & Sex & Weight & Height & HW & LVMass & AFib & VTVF & Diabetes & Hypertension & LVEF\\\\\n",
       "  & <chr> & <chr> & <fct> & <dbl> & <dbl> & <dbl> & <chr> & <fct> & <fct> & <dbl> & <fct> & <dbl> & <dbl> & <int> & <int> & <fct> & <fct> & <fct> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\tC00039 & SRR10676821 & SRX7354049 & Magnet\\_10 & 72.76478 & 8.4 & 0.321948 & NF & NFD & AA & 18 & M & 70 & 175 &  NA & NA & No & No & No & No & 0.37\\\\\n",
       "\tC00055 & SRR10676822 & SRX7354050 & Magnet\\_11 & 74.29391 & 9.1 & 0.461090 & NF & NFD & AA & 26 & M & 85 & 183 &  NA & NA & No & No & No & No &   NA\\\\\n",
       "\tC00074 & SRR10676823 & SRX7354051 & Magnet\\_09 & 77.44517 & 7.8 & 0.338003 & NF & NFD & AA & 17 & M & 68 & 173 & 400 & NA & No & No & No & No & 0.27\\\\\n",
       "\tC00085 & SRR10676824 & SRX7354319 & Magnet\\_06 & 69.51659 & 9.4 & 0.420441 & NF & NFD & AA & 59 & F & 66 & 157 & 380 & NA & No & No & No & No & 0.60\\\\\n",
       "\tC00105 & SRR10676825 & SRX7354320 & Magnet\\_03 & 73.60743 & 8.8 & 0.290037 & NF & NFD & AA & 59 & F & 87 & 157 &  NA & NA & No & No & No & No & 0.55\\\\\n",
       "\tC00120 & SRR10676826 & SRX7354321 & Magnet\\_12 & 30.35077 & 8.6 & 0.595905 & NF & NFD & C  & 50 & M & 87 & 175 & 640 & NA & No & No & No & No & 0.65\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 x 20\n",
       "\n",
       "| <!--/--> | Run &lt;chr&gt; | Experiment &lt;chr&gt; | LibraryPool &lt;fct&gt; | TIN &lt;dbl&gt; | RIN &lt;dbl&gt; | DuplicationRate &lt;dbl&gt; | TissueSource &lt;chr&gt; | Etiology &lt;fct&gt; | Race &lt;fct&gt; | Age &lt;dbl&gt; | Sex &lt;fct&gt; | Weight &lt;dbl&gt; | Height &lt;dbl&gt; | HW &lt;int&gt; | LVMass &lt;int&gt; | AFib &lt;fct&gt; | VTVF &lt;fct&gt; | Diabetes &lt;fct&gt; | Hypertension &lt;fct&gt; | LVEF &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| C00039 | SRR10676821 | SRX7354049 | Magnet_10 | 72.76478 | 8.4 | 0.321948 | NF | NFD | AA | 18 | M | 70 | 175 |  NA | NA | No | No | No | No | 0.37 |\n",
       "| C00055 | SRR10676822 | SRX7354050 | Magnet_11 | 74.29391 | 9.1 | 0.461090 | NF | NFD | AA | 26 | M | 85 | 183 |  NA | NA | No | No | No | No |   NA |\n",
       "| C00074 | SRR10676823 | SRX7354051 | Magnet_09 | 77.44517 | 7.8 | 0.338003 | NF | NFD | AA | 17 | M | 68 | 173 | 400 | NA | No | No | No | No | 0.27 |\n",
       "| C00085 | SRR10676824 | SRX7354319 | Magnet_06 | 69.51659 | 9.4 | 0.420441 | NF | NFD | AA | 59 | F | 66 | 157 | 380 | NA | No | No | No | No | 0.60 |\n",
       "| C00105 | SRR10676825 | SRX7354320 | Magnet_03 | 73.60743 | 8.8 | 0.290037 | NF | NFD | AA | 59 | F | 87 | 157 |  NA | NA | No | No | No | No | 0.55 |\n",
       "| C00120 | SRR10676826 | SRX7354321 | Magnet_12 | 30.35077 | 8.6 | 0.595905 | NF | NFD | C  | 50 | M | 87 | 175 | 640 | NA | No | No | No | No | 0.65 |\n",
       "\n"
      ],
      "text/plain": [
       "       Run         Experiment LibraryPool TIN      RIN DuplicationRate\n",
       "C00039 SRR10676821 SRX7354049 Magnet_10   72.76478 8.4 0.321948       \n",
       "C00055 SRR10676822 SRX7354050 Magnet_11   74.29391 9.1 0.461090       \n",
       "C00074 SRR10676823 SRX7354051 Magnet_09   77.44517 7.8 0.338003       \n",
       "C00085 SRR10676824 SRX7354319 Magnet_06   69.51659 9.4 0.420441       \n",
       "C00105 SRR10676825 SRX7354320 Magnet_03   73.60743 8.8 0.290037       \n",
       "C00120 SRR10676826 SRX7354321 Magnet_12   30.35077 8.6 0.595905       \n",
       "       TissueSource Etiology Race Age Sex Weight Height HW  LVMass AFib VTVF\n",
       "C00039 NF           NFD      AA   18  M   70     175     NA NA     No   No  \n",
       "C00055 NF           NFD      AA   26  M   85     183     NA NA     No   No  \n",
       "C00074 NF           NFD      AA   17  M   68     173    400 NA     No   No  \n",
       "C00085 NF           NFD      AA   59  F   66     157    380 NA     No   No  \n",
       "C00105 NF           NFD      AA   59  F   87     157     NA NA     No   No  \n",
       "C00120 NF           NFD      C    50  M   87     175    640 NA     No   No  \n",
       "       Diabetes Hypertension LVEF\n",
       "C00039 No       No           0.37\n",
       "C00055 No       No             NA\n",
       "C00074 No       No           0.27\n",
       "C00085 No       No           0.60\n",
       "C00105 No       No           0.55\n",
       "C00120 No       No           0.65"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>41886</li><li>354</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 41886\n",
       "\\item 354\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 41886\n",
       "2. 354\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 41886   354"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(colData)\n",
    "all.equal(colnames(countData), rownames(colData))\n",
    "dim(countData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We try to see how the inferred surrogate variables correlate with the covariates that we did not include in the model..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(sva)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "findDeseqFactorsSingle <- function(count_data)\n",
    "{\n",
    "  loggeomeans <- rowMeans(log(count_data))\n",
    "  deseqFactors <- apply(count_data, 2, deseq, loggeomeans = loggeomeans) \n",
    "  deseqFactors\n",
    "}\n",
    "\n",
    "deseq <- function(x, loggeomeans) {\n",
    "  finitePositive <- is.finite(loggeomeans) & x > 0\n",
    "  if (any(finitePositive)) {\n",
    "    res <- exp(median((log(x) - loggeomeans)[finitePositive], na.rm = TRUE))\n",
    "  } else {\n",
    "    print(utils::head(x))\n",
    "    stop(\"Can't normalise accross a condition.\n",
    "         Too many zero expressed genes. \")\n",
    "  }\n",
    "  res\n",
    "}\n",
    "\n",
    "sizeFactors <- findDeseqFactorsSingle(countData)\n",
    "normData <- t(t(countData) / findDeseqFactorsSingle(countData))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# full model\n",
    "# including both the adjustment variables and the variable of interest (here we assume none)\n",
    "# ~Etiology+LibraryPool+TIN+RIN+DuplicationRate+Race+Age+Sex+Weight+Height+HW+AFib+VTVF+Diabetes+Hypertension+LVEF\n",
    "mod = model.matrix(~Etiology, data=colData)\n",
    "\n",
    "# null model\n",
    "# only the adjustment variables\n",
    "mod0 <- model.matrix(~1,data=colData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of significant surrogate variables is:  5 \n",
      "Iteration (out of 5 ):1  2  3  4  5  "
     ]
    }
   ],
   "source": [
    "# estimate number of latent factors\n",
    "# n.sv = num.sv(normData, mod, method=\"leek\")\n",
    "\n",
    "# skip and fix for now to be the number of real adjustment variables (drop LVMass)\n",
    "n.sv <- 5\n",
    "\n",
    "# fix number of surrogate variables...\n",
    "svseq <- svaseq(normData, mod, mod0, n.sv = n.sv)\n",
    "\n",
    "# sv is a matrix whose columns correspond to the estimated surrogate variables (can be used in downstream analyses)\n",
    "# pprob.gam is the posterior probability that each gene is associated with one or more latent variables\n",
    "# pprob.b is the posterior probability that each gene is associated with the variables of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Because we actually do know the real variables (assuming there are no other sources of variation), \n",
    "# we can see how well the SVA method did at recovering these variables "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for (n in c('LibraryPool', 'TIN', 'RIN', 'DuplicationRate', 'Race', 'Age', 'Sex', 'Weight', 'Height', 'HW', 'AFib', 'VTVF', 'Diabetes', 'Hypertension', 'LVEF')){\n",
    "   png(file.path(dirloc, paste(n, 'SVA5.png', sep='-')), width=600, height=350)\n",
    "    par(mfrow = c(2, 3), mar = c(3,5,3,1))\n",
    "    for (i in 1:n.sv) {\n",
    "      stripchart(svseq$sv[, i] ~ colData[[n]], vertical = TRUE, main = paste0(\"SV\", i))\n",
    "      abline(h = 0)\n",
    "     }\n",
    "   dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PVCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# based on https://bioconductor.statistik.tu-dortmund.de/packages/3.7/workflows/vignettes/ExpressionNormalizationWorkflow\n",
    "\n",
    "library(Biobase)\n",
    "library(limma)\n",
    "library(lme4)\n",
    "library(matrixStats)\n",
    "library(pvca)\n",
    "library(vsn)\n",
    "\n",
    "expSetobj <- function(expression, covariates, cov_desc=NULL)\n",
    "{\n",
    "    expression <- as.matrix(expression)\n",
    "    if(is.null(cov_desc)) {\n",
    "        cov_desc <- colnames(covariates)\n",
    "    }\n",
    "    metadata <- data.frame(labelDescription=cov_desc, \n",
    "                           row.names=colnames(covariates))\n",
    "    phenoData <- new(\"AnnotatedDataFrame\", data=covariates,\n",
    "                     varMetadata=metadata)\n",
    "    exp_datObj <- ExpressionSet(assayData=expression, phenoData=phenoData)\n",
    "    return(exp_datObj)\n",
    "}\n",
    "\n",
    "\n",
    "## PVCA by Pierre R.Bushel and Jianying Li\n",
    "pvcaBatchAssess <- function (abatch, batch_factors, threshold)\n",
    "{\n",
    "    theDataMatrix <- exprs(vsn2(abatch, verbose=FALSE))\n",
    "    dataRowN <- nrow(theDataMatrix)\n",
    "    dataColN <- ncol(theDataMatrix)\n",
    "    theDataMatrixCentered <- matrix(data=0, nrow=dataRowN,\n",
    "                                    ncol=dataColN)\n",
    "    theDataMatrixCentered_transposed <- apply(theDataMatrix, 1,\n",
    "                                             scale, center=TRUE, scale=FALSE)\n",
    "    theDataMatrixCentered <- t(theDataMatrixCentered_transposed)\n",
    "    theDataCor <- cor(theDataMatrixCentered)\n",
    "    eigenData <- eigen(theDataCor)\n",
    "    eigenValues <- eigenData$values\n",
    "    ev_n <- length(eigenValues)\n",
    "    eigenVectorsMatrix <- eigenData$vectors\n",
    "    eigenValuesSum <- sum(eigenValues)\n",
    "    percents_PCs <- eigenValues / eigenValuesSum\n",
    "    expInfo <- pData(abatch)[, batch_factors]\n",
    "    exp_design <- as.data.frame(expInfo)\n",
    "    expDesignRowN <- nrow(exp_design)\n",
    "    expDesignColN <- ncol(exp_design)\n",
    "    my_counter_2 <- 0\n",
    "    my_sum_2 <- 1\n",
    "    for (i in ev_n:1) {\n",
    "        my_sum_2 <- my_sum_2 - percents_PCs[i]\n",
    "        if ( (my_sum_2) <= threshold) {\n",
    "            my_counter_2 <- my_counter_2 + 1\n",
    "        }\n",
    "    }\n",
    "    if (my_counter_2 < 3) {\n",
    "        pc_n <- 3\n",
    "    }\n",
    "    else {\n",
    "        pc_n <- my_counter_2\n",
    "    }\n",
    "    pc_data_matrix <- matrix(data=0, nrow= (expDesignRowN *\n",
    "                                                   pc_n), ncol=1)\n",
    "    mycounter <- 0\n",
    "    for (i in 1:pc_n) {\n",
    "        for (j in 1:expDesignRowN) {\n",
    "            mycounter <- mycounter + 1\n",
    "            pc_data_matrix[mycounter, 1] <- eigenVectorsMatrix[j,i]\n",
    "        }\n",
    "    }\n",
    "    AAA <- exp_design[rep(1:expDesignRowN, pc_n), ]\n",
    "    Data <- cbind(AAA, pc_data_matrix)\n",
    "    variables <- c(colnames(exp_design))\n",
    "    for (i in 1:length(variables)) {\n",
    "        Data$variables[i] <- as.factor(Data$variables[i])\n",
    "    }\n",
    "    op <- options(warn= (-1))\n",
    "    effects_n <- expDesignColN + 1  \n",
    "    randomEffectsMatrix <- matrix(data=0, nrow=pc_n, ncol=effects_n)\n",
    "    model.func <- c()\n",
    "    index <- 1\n",
    "    for (i in 1:length(variables)) {\n",
    "        mod <- paste(\"(1|\", variables[i], \")\", sep=\"\")\n",
    "        model.func[index] <- mod\n",
    "        index <- index + 1\n",
    "    }\n",
    "    function.mods <- paste(model.func, collapse=\" + \")\n",
    "    for (i in 1:pc_n) {\n",
    "        y <- (((i - 1) * expDesignRowN) + 1)\n",
    "        funct <- paste(\"pc_data_matrix\", function.mods, sep=\" ~ \")\n",
    "        Rm1ML <- lmer(funct, Data[y: ( ( (i - 1) * expDesignRowN) +\n",
    "                     expDesignRowN), ], REML=TRUE, verbose=FALSE,\n",
    "                            na.action=na.omit)\n",
    "        randomEffects <- Rm1ML\n",
    "        randomEffectsMatrix[i, ] <- c(unlist(VarCorr(Rm1ML)),\n",
    "                                      resid=sigma(Rm1ML)^2)\n",
    "    }\n",
    "    effectsNames <- c(names(getME(Rm1ML, \"cnms\")), \"resid\")\n",
    "    randomEffectsMatrixStdze <- matrix(data=0, nrow=pc_n,\n",
    "                                       ncol=effects_n)\n",
    "    for (i in 1:pc_n) {\n",
    "        mySum <- sum(randomEffectsMatrix[i, ])\n",
    "        for (j in 1:effects_n) {\n",
    "            randomEffectsMatrixStdze[i, j] <- randomEffectsMatrix[i,\n",
    "                                                                 j] / mySum\n",
    "        }\n",
    "    }\n",
    "    randomEffectsMatrixWtProp <- matrix(data=0, nrow=pc_n,\n",
    "                                        ncol=effects_n)\n",
    "    for (i in 1:pc_n) {\n",
    "        weight <- eigenValues[i] / eigenValuesSum\n",
    "        for (j in 1:effects_n) {\n",
    "            randomEffectsMatrixWtProp[i, j] <- randomEffectsMatrixStdze[i,j] * weight\n",
    "        }\n",
    "    }\n",
    "    randomEffectsSums <- matrix(data=0, nrow=1, ncol=effects_n)\n",
    "    randomEffectsSums <- colSums(randomEffectsMatrixWtProp)\n",
    "    totalSum <- sum(randomEffectsSums)\n",
    "    randomEffectsMatrixWtAveProp <- matrix(data=0, nrow=1,\n",
    "                                           ncol=effects_n)\n",
    "    for (j in 1:effects_n) {\n",
    "        randomEffectsMatrixWtAveProp[j] <- randomEffectsSums[j] / totalSum\n",
    "    }\n",
    "    return(list(dat=randomEffectsMatrixWtAveProp, label=effectsNames))\n",
    "}\n",
    "\n",
    "snmAnaly <- function(exprs, cov, bv, av, iv, adj=TRUE)\n",
    "{\n",
    "    kar_rawdata <- exprs\n",
    "    kar_bio.var <- as.data.frame(cov[,bv])\n",
    "    kar_adj.var <- as.data.frame(cov[,av])\n",
    "    rownames(kar_adj.var) <- rownames(cov)\n",
    "    colnames(kar_adj.var) <- av\n",
    "    kar_int.var <- as.data.frame(cov[,iv])\n",
    "    colnames(kar_int.var) <- iv\n",
    "    int.var <- kar_int.var\n",
    "    int.var$Array <- as.factor(int.var$Array)\n",
    "    adj.var <- model.matrix(~., kar_adj.var)\n",
    "    bio.var <- model.matrix(~., kar_bio.var)\n",
    "    raw.data <- as.matrix(kar_rawdata)\n",
    "    snm.obj <- snm(raw.data, bio.var, adj.var, int.var, rm.adj=adj, num.iter=5)\n",
    "    write.table(snm.obj$norm.dat, file=\"snm_normalized_data.csv\", \n",
    "                sep=\",\", col.names=NA)\n",
    "    return(snm.obj)\n",
    "}\n",
    "\n",
    "\n",
    "conTocat <- function(covariates, var_names)\n",
    "{\n",
    "    var <- covariates[,var_names]\n",
    "    len <- dim(var)[2]\n",
    "    for (i in 1:len) {\n",
    "        temp <- as.numeric(with(var, cut(var[,i], breaks=\n",
    "                          quantile(var[,i], probs=seq(0, 1, by=0.25)), \n",
    "                                                  include.lowest=TRUE)))\n",
    "        covariates <- cbind(covariates, temp)\n",
    "        colnames(covariates)[ncol(covariates)] <- \n",
    "                           paste(var_names[i], \"_cat\", sep=\"\")\n",
    "    }\n",
    "    return(covariates)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PVCA (Bushel, 2013) estimates the proportion of the variance of the principal components of the gene expression data that can be attributed to each of the given covariates.The remaining fraction is assigned as “residual” variance.It efficiently combines principal component analysis (PCA) to reduce the feature space and variance component analysis (VCA) which fits a mixed linear model using factors of interest as random effects to estimate and partition the total variability. The variance explained is computed as a weighted average of the contributions of each factor to each PC, and you have the option of specifying how many principal components to include, by setting a threshold amount of the variance that needs to be explained by the identified PCs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we first need to categorize the continuous covariates\n",
    "\n",
    "colData$TIN <- cut(colData$TIN, \n",
    "                   breaks = quantile(colData$TIN, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$RIN <- cut(colData$RIN, \n",
    "                   breaks = quantile(colData$RIN, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$DuplicationRate <- cut(colData$DuplicationRate, \n",
    "                   breaks = quantile(colData$DuplicationRate, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$Age <- cut(colData$Age, \n",
    "                   breaks = quantile(colData$Age, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$Weight <- cut(colData$Weight, \n",
    "                   breaks = quantile(colData$Weight, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$Height <- cut(colData$Height, \n",
    "                   breaks = quantile(colData$Height, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$HW <- cut(colData$HW, \n",
    "                   breaks = quantile(colData$HW, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)\n",
    "colData$LVEF <- cut(colData$LVEF, \n",
    "                   breaks = quantile(colData$LVEF, by=0.25, na.rm = TRUE), \n",
    "                   include.lowest = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# and remove unused variables\n",
    "colData <- colData[,-c(1,2,7,15)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 16</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>LibraryPool</th><th scope=col>TIN</th><th scope=col>RIN</th><th scope=col>DuplicationRate</th><th scope=col>Etiology</th><th scope=col>Race</th><th scope=col>Age</th><th scope=col>Sex</th><th scope=col>Weight</th><th scope=col>Height</th><th scope=col>HW</th><th scope=col>AFib</th><th scope=col>VTVF</th><th scope=col>Diabetes</th><th scope=col>Hypertension</th><th scope=col>LVEF</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>C00039</th><td>Magnet_10</td><td>(70.1,73.3]</td><td>(8.1,8.5]</td><td>[0.233,0.378]</td><td>NFD</td><td>AA</td><td>[15,48]</td><td>M</td><td>(66.2,79]</td><td>(170,178]</td><td>NA       </td><td>No</td><td>No</td><td>No</td><td>No</td><td>(0.2,0.52]</td></tr>\n",
       "\t<tr><th scope=row>C00055</th><td>Magnet_11</td><td>(73.3,81.5]</td><td>(8.9,10] </td><td>(0.421,0.517]</td><td>NFD</td><td>AA</td><td>[15,48]</td><td>M</td><td>(79,91]  </td><td>(178,196]</td><td>NA       </td><td>No</td><td>No</td><td>No</td><td>No</td><td>NA        </td></tr>\n",
       "\t<tr><th scope=row>C00074</th><td>Magnet_09</td><td>(73.3,81.5]</td><td>[5.8,8.1]</td><td>[0.233,0.378]</td><td>NFD</td><td>AA</td><td>[15,48]</td><td>M</td><td>(66.2,79]</td><td>(170,178]</td><td>(372,448]</td><td>No</td><td>No</td><td>No</td><td>No</td><td>(0.2,0.52]</td></tr>\n",
       "\t<tr><th scope=row>C00085</th><td>Magnet_06</td><td>(60.6,70.1]</td><td>(8.9,10] </td><td>(0.378,0.421]</td><td>NFD</td><td>AA</td><td>(56,62]</td><td>F</td><td>[27,66.2]</td><td>[55,163] </td><td>(372,448]</td><td>No</td><td>No</td><td>No</td><td>No</td><td>(0.52,0.8]</td></tr>\n",
       "\t<tr><th scope=row>C00105</th><td>Magnet_03</td><td>(73.3,81.5]</td><td>(8.5,8.9]</td><td>[0.233,0.378]</td><td>NFD</td><td>AA</td><td>(56,62]</td><td>F</td><td>(79,91]  </td><td>[55,163] </td><td>NA       </td><td>No</td><td>No</td><td>No</td><td>No</td><td>(0.52,0.8]</td></tr>\n",
       "\t<tr><th scope=row>C00120</th><td>Magnet_12</td><td>[23.4,60.6]</td><td>(8.5,8.9]</td><td>(0.517,0.935]</td><td>NFD</td><td>C </td><td>(48,56]</td><td>M</td><td>(79,91]  </td><td>(170,178]</td><td>(558,923]</td><td>No</td><td>No</td><td>No</td><td>No</td><td>(0.52,0.8]</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 x 16\n",
       "\\begin{tabular}{r|llllllllllllllll}\n",
       "  & LibraryPool & TIN & RIN & DuplicationRate & Etiology & Race & Age & Sex & Weight & Height & HW & AFib & VTVF & Diabetes & Hypertension & LVEF\\\\\n",
       "  & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\tC00039 & Magnet\\_10 & (70.1,73.3{]} & (8.1,8.5{]} & {[}0.233,0.378{]} & NFD & AA & {[}15,48{]} & M & (66.2,79{]} & (170,178{]} & NA        & No & No & No & No & (0.2,0.52{]}\\\\\n",
       "\tC00055 & Magnet\\_11 & (73.3,81.5{]} & (8.9,10{]}  & (0.421,0.517{]} & NFD & AA & {[}15,48{]} & M & (79,91{]}   & (178,196{]} & NA        & No & No & No & No & NA        \\\\\n",
       "\tC00074 & Magnet\\_09 & (73.3,81.5{]} & {[}5.8,8.1{]} & {[}0.233,0.378{]} & NFD & AA & {[}15,48{]} & M & (66.2,79{]} & (170,178{]} & (372,448{]} & No & No & No & No & (0.2,0.52{]}\\\\\n",
       "\tC00085 & Magnet\\_06 & (60.6,70.1{]} & (8.9,10{]}  & (0.378,0.421{]} & NFD & AA & (56,62{]} & F & {[}27,66.2{]} & {[}55,163{]}  & (372,448{]} & No & No & No & No & (0.52,0.8{]}\\\\\n",
       "\tC00105 & Magnet\\_03 & (73.3,81.5{]} & (8.5,8.9{]} & {[}0.233,0.378{]} & NFD & AA & (56,62{]} & F & (79,91{]}   & {[}55,163{]}  & NA        & No & No & No & No & (0.52,0.8{]}\\\\\n",
       "\tC00120 & Magnet\\_12 & {[}23.4,60.6{]} & (8.5,8.9{]} & (0.517,0.935{]} & NFD & C  & (48,56{]} & M & (79,91{]}   & (170,178{]} & (558,923{]} & No & No & No & No & (0.52,0.8{]}\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 x 16\n",
       "\n",
       "| <!--/--> | LibraryPool &lt;fct&gt; | TIN &lt;fct&gt; | RIN &lt;fct&gt; | DuplicationRate &lt;fct&gt; | Etiology &lt;fct&gt; | Race &lt;fct&gt; | Age &lt;fct&gt; | Sex &lt;fct&gt; | Weight &lt;fct&gt; | Height &lt;fct&gt; | HW &lt;fct&gt; | AFib &lt;fct&gt; | VTVF &lt;fct&gt; | Diabetes &lt;fct&gt; | Hypertension &lt;fct&gt; | LVEF &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| C00039 | Magnet_10 | (70.1,73.3] | (8.1,8.5] | [0.233,0.378] | NFD | AA | [15,48] | M | (66.2,79] | (170,178] | NA        | No | No | No | No | (0.2,0.52] |\n",
       "| C00055 | Magnet_11 | (73.3,81.5] | (8.9,10]  | (0.421,0.517] | NFD | AA | [15,48] | M | (79,91]   | (178,196] | NA        | No | No | No | No | NA         |\n",
       "| C00074 | Magnet_09 | (73.3,81.5] | [5.8,8.1] | [0.233,0.378] | NFD | AA | [15,48] | M | (66.2,79] | (170,178] | (372,448] | No | No | No | No | (0.2,0.52] |\n",
       "| C00085 | Magnet_06 | (60.6,70.1] | (8.9,10]  | (0.378,0.421] | NFD | AA | (56,62] | F | [27,66.2] | [55,163]  | (372,448] | No | No | No | No | (0.52,0.8] |\n",
       "| C00105 | Magnet_03 | (73.3,81.5] | (8.5,8.9] | [0.233,0.378] | NFD | AA | (56,62] | F | (79,91]   | [55,163]  | NA        | No | No | No | No | (0.52,0.8] |\n",
       "| C00120 | Magnet_12 | [23.4,60.6] | (8.5,8.9] | (0.517,0.935] | NFD | C  | (48,56] | M | (79,91]   | (170,178] | (558,923] | No | No | No | No | (0.52,0.8] |\n",
       "\n"
      ],
      "text/plain": [
       "       LibraryPool TIN         RIN       DuplicationRate Etiology Race Age    \n",
       "C00039 Magnet_10   (70.1,73.3] (8.1,8.5] [0.233,0.378]   NFD      AA   [15,48]\n",
       "C00055 Magnet_11   (73.3,81.5] (8.9,10]  (0.421,0.517]   NFD      AA   [15,48]\n",
       "C00074 Magnet_09   (73.3,81.5] [5.8,8.1] [0.233,0.378]   NFD      AA   [15,48]\n",
       "C00085 Magnet_06   (60.6,70.1] (8.9,10]  (0.378,0.421]   NFD      AA   (56,62]\n",
       "C00105 Magnet_03   (73.3,81.5] (8.5,8.9] [0.233,0.378]   NFD      AA   (56,62]\n",
       "C00120 Magnet_12   [23.4,60.6] (8.5,8.9] (0.517,0.935]   NFD      C    (48,56]\n",
       "       Sex Weight    Height    HW        AFib VTVF Diabetes Hypertension\n",
       "C00039 M   (66.2,79] (170,178] NA        No   No   No       No          \n",
       "C00055 M   (79,91]   (178,196] NA        No   No   No       No          \n",
       "C00074 M   (66.2,79] (170,178] (372,448] No   No   No       No          \n",
       "C00085 F   [27,66.2] [55,163]  (372,448] No   No   No       No          \n",
       "C00105 F   (79,91]   [55,163]  NA        No   No   No       No          \n",
       "C00120 M   (79,91]   (170,178] (558,923] No   No   No       No          \n",
       "       LVEF      \n",
       "C00039 (0.2,0.52]\n",
       "C00055 NA        \n",
       "C00074 (0.2,0.52]\n",
       "C00085 (0.52,0.8]\n",
       "C00105 (0.52,0.8]\n",
       "C00120 (0.52,0.8]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(colData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert to ExpressionSet\n",
    "# inpData <- expSetobj(countData, colData)\n",
    "\n",
    "# use norm - uncomment otherwise\n",
    "sizeFactors <- findDeseqFactorsSingle(countData)\n",
    "normData <- t(t(countData) / findDeseqFactorsSingle(countData))\n",
    "inpData <- expSetobj(normData, colData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# set the covariates whose effect size on the data needs to be calculated\n",
    "cvrts_eff_var <- colnames(colData)\n",
    "## PVCA Threshold Value is the percentile value of the minimum amount of the variabilities that the selected principal components need to explain, \n",
    "## here requiring 75% of the expression variance to be captured by the PCs\n",
    "pct_thrsh <- 0.75 \n",
    "## Perform the PVCA analysis\n",
    "pvcaObj <- pvcaBatchAssess(inpData, cvrts_eff_var, pct_thrsh)\n",
    "\n",
    "# note that the model is singular...\n",
    "# we could try reducing the number of covariates, etc.? -> there is no straighforward solution though..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot\n",
    "\n",
    "png(file.path(dirloc, 'PVCA_norm.png'), width=800, height=400)\n",
    "bp <- barplot(pvcaObj$dat,\n",
    "              ylab=expression(bold(\"Weighted Average Proportion Variance\")),\n",
    "              ylim=c(0,1.1), col=c(\"navy\"), las=2, font=4,\n",
    "              main=\"Principal Variant Component Analysis Estimation\")\n",
    "axis(1,at=bp,labels=pvcaObj$label,cex.axis=1,las=2,font=4)\n",
    "values <- pvcaObj$dat\n",
    "new_values <- round(values,3)\n",
    "text(bp,pvcaObj$dat,labels=new_values,pos=3,cex=1,font=4)\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(pvcaObj, file.path(dirloc, 'pvcaObj.RDS'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Surrogate Variable Analysis ...\n",
    "\n",
    "# These are appended as new covariates to the existing list of covariates, and \n",
    "# we also add a step to convert them to categorical variables so as to estimate their contributions to the total variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of significant surrogate variables is:  5 \n",
      "Iteration (out of 5 ):1  2  3  4  5  "
     ]
    }
   ],
   "source": [
    "# choose a biological variable that is to be used to calculate the surrogate variables\n",
    "biol_var_sva <- \"Etiology\" \n",
    "# Perform SVA as above\n",
    "pheno <- pData(inpData)\n",
    "expData <- exprs(inpData) # use norm\n",
    "mod  <- model.matrix(~as.factor(pheno[, biol_var_sva]), data=pheno)\n",
    "mod0 <- model.matrix(~1,data=pheno)\n",
    "\n",
    "n.sv <- 5\n",
    "svobj <- svaseq(expData, mod, mod0, n.sv=n.sv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat\n",
    "srg_var <- as.data.frame(svobj$sv)     \n",
    "colnames(srg_var) <- paste(\"sv\", 1:ncol(srg_var), sep=\"\")\n",
    "pData(inpData) <- cbind(pData(inpData), srg_var)\n",
    "varMetadata(inpData)[colnames(srg_var),] <- colnames(srg_var)\n",
    "svobj$expSetobject <- inpData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## append to the ExpressionSet object\n",
    "inpData_sv <- svobj$expSetobject"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Computing the correlation between the surrogate variables and the covariates...\n",
    "\n",
    "We want to identify if a given surrogate variable is entirely independent of all the existing covariates or if it correlates with some technical factor, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = pData(inpData_sv)[, \"sv1\"] ~ pData(inpData_sv)[, \n",
       "    \"LibraryPool\"] + pData(inpData_sv)[, \"TIN\"] + pData(inpData_sv)[, \n",
       "    \"RIN\"] + pData(inpData_sv)[, \"DuplicationRate\"] + pData(inpData_sv)[, \n",
       "    \"Etiology\"] + pData(inpData_sv)[, \"Race\"] + pData(inpData_sv)[, \n",
       "    \"Sex\"] + pData(inpData_sv)[, \"Weight\"] + pData(inpData_sv)[, \n",
       "    \"Height\"] + pData(inpData_sv)[, \"HW\"] + pData(inpData_sv)[, \n",
       "    \"AFib\"] + pData(inpData_sv)[, \"VTVF\"] + pData(inpData_sv)[, \n",
       "    \"Diabetes\"] + pData(inpData_sv)[, \"Hypertension\"] + pData(inpData_sv)[, \n",
       "    \"LVEF\"])\n",
       "\n",
       "Deviance Residuals: \n",
       "      Min         1Q     Median         3Q        Max  \n",
       "-0.068446  -0.011751  -0.000268   0.010948   0.091576  \n",
       "\n",
       "Coefficients: (1 not defined because of singularities)\n",
       "                                                      Estimate Std. Error\n",
       "(Intercept)                                         -0.0544406  0.0121796\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02         -0.0213014  0.0081037\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03          0.0018358  0.0074480\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04         -0.0040595  0.0075208\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05          0.0122394  0.0078443\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06          0.0102433  0.0084503\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07          0.0065129  0.0074802\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08          0.0128071  0.0071579\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09         -0.0256426  0.0079350\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10          0.0004229  0.0077450\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11         -0.0161471  0.0077636\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12         -0.0441670  0.0075925\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                0.0679096  0.0051705\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                0.0823595  0.0058423\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                0.0878010  0.0059488\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                  0.0046178  0.0043192\n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                  0.0067870  0.0042704\n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                   0.0082079  0.0050322\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421] -0.0029964  0.0045108\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517] -0.0023253  0.0055233\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935] -0.0152526  0.0068285\n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                  -0.0011696  0.0063048\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                   0.0014860  0.0071207\n",
       "pData(inpData_sv)[, \"Race\"]AA                        0.0037931  0.0033875\n",
       "pData(inpData_sv)[, \"Sex\"]F                          0.0031177  0.0050830\n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]              -0.0003987  0.0041544\n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                -0.0031303  0.0045727\n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                0.0033427  0.0049485\n",
       "pData(inpData_sv)[, \"Height\"](163,170]               0.0092018  0.0044619\n",
       "pData(inpData_sv)[, \"Height\"](170,178]               0.0071917  0.0057907\n",
       "pData(inpData_sv)[, \"Height\"](178,196]               0.0069474  0.0062068\n",
       "pData(inpData_sv)[, \"HW\"](372,448]                  -0.0041579  0.0044490\n",
       "pData(inpData_sv)[, \"HW\"](448,558]                  -0.0070278  0.0046239\n",
       "pData(inpData_sv)[, \"HW\"](558,923]                  -0.0063997  0.0048380\n",
       "pData(inpData_sv)[, \"AFib\"]Yes                       0.0002670  0.0036085\n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                       0.0026320  0.0037205\n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                   0.0004194  0.0037053\n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                      NA         NA\n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]               -0.0039887  0.0047020\n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]               -0.0059407  0.0050952\n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]               -0.0011782  0.0067976\n",
       "                                                    t value Pr(>|t|)    \n",
       "(Intercept)                                          -4.470 1.25e-05 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02          -2.629  0.00917 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03           0.246  0.80554    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04          -0.540  0.58989    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05           1.560  0.12011    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06           1.212  0.22673    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07           0.871  0.38487    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08           1.789  0.07494 .  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09          -3.232  0.00142 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10           0.055  0.95651    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11          -2.080  0.03869 *  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12          -5.817 2.07e-08 ***\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                13.134  < 2e-16 ***\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                14.097  < 2e-16 ***\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                14.760  < 2e-16 ***\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                   1.069  0.28617    \n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                   1.589  0.11341    \n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                    1.631  0.10429    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421]  -0.664  0.50721    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517]  -0.421  0.67416    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935]  -2.234  0.02650 *  \n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                   -0.186  0.85300    \n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                    0.209  0.83488    \n",
       "pData(inpData_sv)[, \"Race\"]AA                         1.120  0.26404    \n",
       "pData(inpData_sv)[, \"Sex\"]F                           0.613  0.54027    \n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]               -0.096  0.92363    \n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                 -0.685  0.49433    \n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                 0.675  0.50006    \n",
       "pData(inpData_sv)[, \"Height\"](163,170]                2.062  0.04034 *  \n",
       "pData(inpData_sv)[, \"Height\"](170,178]                1.242  0.21557    \n",
       "pData(inpData_sv)[, \"Height\"](178,196]                1.119  0.26422    \n",
       "pData(inpData_sv)[, \"HW\"](372,448]                   -0.935  0.35102    \n",
       "pData(inpData_sv)[, \"HW\"](448,558]                   -1.520  0.12996    \n",
       "pData(inpData_sv)[, \"HW\"](558,923]                   -1.323  0.18727    \n",
       "pData(inpData_sv)[, \"AFib\"]Yes                        0.074  0.94109    \n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                        0.707  0.48004    \n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                    0.113  0.90997    \n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                   NA       NA    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                -0.848  0.39719    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                -1.166  0.24489    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                -0.173  0.86255    \n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for gaussian family taken to be 0.000497756)\n",
       "\n",
       "    Null deviance: 0.78249  on 261  degrees of freedom\n",
       "Residual deviance: 0.11050  on 222  degrees of freedom\n",
       "  (92 observations deleted due to missingness)\n",
       "AIC: -1210.5\n",
       "\n",
       "Number of Fisher Scoring iterations: 2\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Fit a generalized linear model for sv1\n",
    "glm.sv1 <- glm(pData(inpData_sv)[,\"sv1\"] ~ pData(inpData_sv)[,\"LibraryPool\"] + \n",
    "               pData(inpData_sv)[,\"TIN\"] + pData(inpData_sv)[,\"RIN\"] +\n",
    "               pData(inpData_sv)[,\"DuplicationRate\"] + pData(inpData_sv)[,\"Etiology\"] +\n",
    "               pData(inpData_sv)[,\"Race\"] + pData(inpData_sv)[,\"Sex\"] + pData(inpData_sv)[,\"Weight\"] +\n",
    "               pData(inpData_sv)[,\"Height\"] + pData(inpData_sv)[,\"HW\"] + pData(inpData_sv)[,\"AFib\"] +\n",
    "               pData(inpData_sv)[,\"VTVF\"] + pData(inpData_sv)[,\"Diabetes\"] + pData(inpData_sv)[,\"Hypertension\"] +\n",
    "               pData(inpData_sv)[,\"LVEF\"]) \n",
    "summary(glm.sv1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = pData(inpData_sv)[, \"sv2\"] ~ pData(inpData_sv)[, \n",
       "    \"LibraryPool\"] + pData(inpData_sv)[, \"TIN\"] + pData(inpData_sv)[, \n",
       "    \"RIN\"] + pData(inpData_sv)[, \"DuplicationRate\"] + pData(inpData_sv)[, \n",
       "    \"Etiology\"] + pData(inpData_sv)[, \"Race\"] + pData(inpData_sv)[, \n",
       "    \"Sex\"] + pData(inpData_sv)[, \"Weight\"] + pData(inpData_sv)[, \n",
       "    \"Height\"] + pData(inpData_sv)[, \"HW\"] + pData(inpData_sv)[, \n",
       "    \"AFib\"] + pData(inpData_sv)[, \"VTVF\"] + pData(inpData_sv)[, \n",
       "    \"Diabetes\"] + pData(inpData_sv)[, \"Hypertension\"] + pData(inpData_sv)[, \n",
       "    \"LVEF\"])\n",
       "\n",
       "Deviance Residuals: \n",
       "      Min         1Q     Median         3Q        Max  \n",
       "-0.144087  -0.015522   0.002433   0.015826   0.097397  \n",
       "\n",
       "Coefficients: (1 not defined because of singularities)\n",
       "                                                      Estimate Std. Error\n",
       "(Intercept)                                          0.0749445  0.0174754\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02         -0.0167360  0.0116272\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03          0.0030001  0.0106864\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04         -0.0425959  0.0107909\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05          0.0219386  0.0112551\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06          0.0052072  0.0121245\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07          0.0165527  0.0107326\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08          0.0248293  0.0102702\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09         -0.0773370  0.0113852\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10          0.0253699  0.0111126\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11         -0.0594414  0.0111392\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12         -0.0068786  0.0108938\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]               -0.0228174  0.0074186\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]               -0.0438373  0.0083825\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]               -0.0665223  0.0085353\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                 -0.0066271  0.0061972\n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                  0.0126404  0.0061272\n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                   0.0059018  0.0072202\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421] -0.0064193  0.0064721\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517] -0.0225989  0.0079249\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935] -0.0108566  0.0097976\n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                  -0.0352754  0.0090462\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                  -0.0205044  0.0102168\n",
       "pData(inpData_sv)[, \"Race\"]AA                        0.0104396  0.0048604\n",
       "pData(inpData_sv)[, \"Sex\"]F                         -0.0037093  0.0072931\n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]               0.0051110  0.0059608\n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                 0.0030688  0.0065609\n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                0.0096655  0.0071002\n",
       "pData(inpData_sv)[, \"Height\"](163,170]               0.0127788  0.0064019\n",
       "pData(inpData_sv)[, \"Height\"](170,178]              -0.0060506  0.0083086\n",
       "pData(inpData_sv)[, \"Height\"](178,196]              -0.0034030  0.0089056\n",
       "pData(inpData_sv)[, \"HW\"](372,448]                  -0.0081659  0.0063834\n",
       "pData(inpData_sv)[, \"HW\"](448,558]                  -0.0149919  0.0066344\n",
       "pData(inpData_sv)[, \"HW\"](558,923]                  -0.0099414  0.0069416\n",
       "pData(inpData_sv)[, \"AFib\"]Yes                      -0.0027719  0.0051776\n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                       0.0121227  0.0053382\n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                   0.0007667  0.0053164\n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                      NA         NA\n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                0.0020634  0.0067465\n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]               -0.0163654  0.0073106\n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]               -0.0160865  0.0097532\n",
       "                                                    t value Pr(>|t|)    \n",
       "(Intercept)                                           4.289 2.68e-05 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02          -1.439 0.151450    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03           0.281 0.779171    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04          -3.947 0.000106 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05           1.949 0.052528 .  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06           0.429 0.667995    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07           1.542 0.124430    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08           2.418 0.016431 *  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09          -6.793 9.96e-11 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10           2.283 0.023378 *  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11          -5.336 2.34e-07 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12          -0.631 0.528411    \n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                -3.076 0.002364 ** \n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                -5.230 3.92e-07 ***\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                -7.794 2.49e-13 ***\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                  -1.069 0.286068    \n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                   2.063 0.040275 *  \n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                    0.817 0.414575    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421]  -0.992 0.322355    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517]  -2.852 0.004760 ** \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935]  -1.108 0.269025    \n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                   -3.899 0.000128 ***\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                   -2.007 0.045969 *  \n",
       "pData(inpData_sv)[, \"Race\"]AA                         2.148 0.032807 *  \n",
       "pData(inpData_sv)[, \"Sex\"]F                          -0.509 0.611528    \n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]                0.857 0.392131    \n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                  0.468 0.640422    \n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                 1.361 0.174797    \n",
       "pData(inpData_sv)[, \"Height\"](163,170]                1.996 0.047147 *  \n",
       "pData(inpData_sv)[, \"Height\"](170,178]               -0.728 0.467237    \n",
       "pData(inpData_sv)[, \"Height\"](178,196]               -0.382 0.702741    \n",
       "pData(inpData_sv)[, \"HW\"](372,448]                   -1.279 0.202148    \n",
       "pData(inpData_sv)[, \"HW\"](448,558]                   -2.260 0.024808 *  \n",
       "pData(inpData_sv)[, \"HW\"](558,923]                   -1.432 0.153508    \n",
       "pData(inpData_sv)[, \"AFib\"]Yes                       -0.535 0.592928    \n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                        2.271 0.024109 *  \n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                    0.144 0.885466    \n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                   NA       NA    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                 0.306 0.760009    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                -2.239 0.026176 *  \n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                -1.649 0.100489    \n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for gaussian family taken to be 0.001024713)\n",
       "\n",
       "    Null deviance: 0.73687  on 261  degrees of freedom\n",
       "Residual deviance: 0.22749  on 222  degrees of freedom\n",
       "  (92 observations deleted due to missingness)\n",
       "AIC: -1021.3\n",
       "\n",
       "Number of Fisher Scoring iterations: 2\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glm.sv2 <- glm(pData(inpData_sv)[,\"sv2\"] ~ pData(inpData_sv)[,\"LibraryPool\"] + \n",
    "               pData(inpData_sv)[,\"TIN\"] + pData(inpData_sv)[,\"RIN\"] +\n",
    "               pData(inpData_sv)[,\"DuplicationRate\"] + pData(inpData_sv)[,\"Etiology\"] +\n",
    "               pData(inpData_sv)[,\"Race\"] + pData(inpData_sv)[,\"Sex\"] + pData(inpData_sv)[,\"Weight\"] +\n",
    "               pData(inpData_sv)[,\"Height\"] + pData(inpData_sv)[,\"HW\"] + pData(inpData_sv)[,\"AFib\"] +\n",
    "               pData(inpData_sv)[,\"VTVF\"] + pData(inpData_sv)[,\"Diabetes\"] + pData(inpData_sv)[,\"Hypertension\"] +\n",
    "               pData(inpData_sv)[,\"LVEF\"]) \n",
    "summary(glm.sv2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = pData(inpData_sv)[, \"sv3\"] ~ pData(inpData_sv)[, \n",
       "    \"LibraryPool\"] + pData(inpData_sv)[, \"TIN\"] + pData(inpData_sv)[, \n",
       "    \"RIN\"] + pData(inpData_sv)[, \"DuplicationRate\"] + pData(inpData_sv)[, \n",
       "    \"Etiology\"] + pData(inpData_sv)[, \"Race\"] + pData(inpData_sv)[, \n",
       "    \"Sex\"] + pData(inpData_sv)[, \"Weight\"] + pData(inpData_sv)[, \n",
       "    \"Height\"] + pData(inpData_sv)[, \"HW\"] + pData(inpData_sv)[, \n",
       "    \"AFib\"] + pData(inpData_sv)[, \"VTVF\"] + pData(inpData_sv)[, \n",
       "    \"Diabetes\"] + pData(inpData_sv)[, \"Hypertension\"] + pData(inpData_sv)[, \n",
       "    \"LVEF\"])\n",
       "\n",
       "Deviance Residuals: \n",
       "      Min         1Q     Median         3Q        Max  \n",
       "-0.176157  -0.028698  -0.003764   0.022816   0.222160  \n",
       "\n",
       "Coefficients: (1 not defined because of singularities)\n",
       "                                                      Estimate Std. Error\n",
       "(Intercept)                                          0.0012833  0.0273612\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02         -0.0625124  0.0182047\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03         -0.0375263  0.0167318\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04         -0.0260171  0.0168954\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05         -0.0108632  0.0176221\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06         -0.0096402  0.0189834\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07         -0.0513004  0.0168041\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08         -0.0183515  0.0160801\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09         -0.0180665  0.0178259\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10         -0.0192992  0.0173991\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11          0.0241449  0.0174408\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12          0.0559118  0.0170565\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                0.0210499  0.0116154\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                0.0168794  0.0131246\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                0.0144386  0.0133638\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                  0.0036272  0.0097030\n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                 -0.0045786  0.0095933\n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                   0.0068773  0.0113048\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421] -0.0082722  0.0101334\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517] -0.0225065  0.0124080\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935] -0.0441025  0.0153401\n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                  -0.0040130  0.0141637\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                  -0.0151798  0.0159964\n",
       "pData(inpData_sv)[, \"Race\"]AA                       -0.0143211  0.0076100\n",
       "pData(inpData_sv)[, \"Sex\"]F                          0.0130848  0.0114188\n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]               0.0070013  0.0093328\n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                 0.0076186  0.0102724\n",
       "pData(inpData_sv)[, \"Weight\"](91,267]               -0.0040590  0.0111167\n",
       "pData(inpData_sv)[, \"Height\"](163,170]              -0.0066753  0.0100235\n",
       "pData(inpData_sv)[, \"Height\"](170,178]              -0.0114051  0.0130088\n",
       "pData(inpData_sv)[, \"Height\"](178,196]               0.0021208  0.0139435\n",
       "pData(inpData_sv)[, \"HW\"](372,448]                   0.0155799  0.0099945\n",
       "pData(inpData_sv)[, \"HW\"](448,558]                   0.0089405  0.0103875\n",
       "pData(inpData_sv)[, \"HW\"](558,923]                   0.0221622  0.0108686\n",
       "pData(inpData_sv)[, \"AFib\"]Yes                       0.0102888  0.0081065\n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                       0.0056010  0.0083580\n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                   0.0121057  0.0083239\n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                      NA         NA\n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                0.0031359  0.0105629\n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                0.0142932  0.0114463\n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                0.0003372  0.0152706\n",
       "                                                    t value Pr(>|t|)    \n",
       "(Intercept)                                           0.047  0.96263    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02          -3.434  0.00071 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03          -2.243  0.02590 *  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04          -1.540  0.12501    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05          -0.616  0.53823    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06          -0.508  0.61208    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07          -3.053  0.00254 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08          -1.141  0.25499    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09          -1.013  0.31193    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10          -1.109  0.26854    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11           1.384  0.16763    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12           3.278  0.00121 ** \n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                 1.812  0.07130 .  \n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                 1.286  0.19975    \n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                 1.080  0.28113    \n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                   0.374  0.70889    \n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                  -0.477  0.63364    \n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                    0.608  0.54358    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421]  -0.816  0.41519    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517]  -1.814  0.07105 .  \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935]  -2.875  0.00443 ** \n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                   -0.283  0.77719    \n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                   -0.949  0.34368    \n",
       "pData(inpData_sv)[, \"Race\"]AA                        -1.882  0.06116 .  \n",
       "pData(inpData_sv)[, \"Sex\"]F                           1.146  0.25307    \n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]                0.750  0.45394    \n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                  0.742  0.45908    \n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                -0.365  0.71536    \n",
       "pData(inpData_sv)[, \"Height\"](163,170]               -0.666  0.50613    \n",
       "pData(inpData_sv)[, \"Height\"](170,178]               -0.877  0.38158    \n",
       "pData(inpData_sv)[, \"Height\"](178,196]                0.152  0.87924    \n",
       "pData(inpData_sv)[, \"HW\"](372,448]                    1.559  0.12046    \n",
       "pData(inpData_sv)[, \"HW\"](448,558]                    0.861  0.39033    \n",
       "pData(inpData_sv)[, \"HW\"](558,923]                    2.039  0.04262 *  \n",
       "pData(inpData_sv)[, \"AFib\"]Yes                        1.269  0.20570    \n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                        0.670  0.50347    \n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                    1.454  0.14727    \n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                   NA       NA    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                 0.297  0.76684    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                 1.249  0.21308    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                 0.022  0.98240    \n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for gaussian family taken to be 0.002512012)\n",
       "\n",
       "    Null deviance: 0.85747  on 261  degrees of freedom\n",
       "Residual deviance: 0.55767  on 222  degrees of freedom\n",
       "  (92 observations deleted due to missingness)\n",
       "AIC: -786.39\n",
       "\n",
       "Number of Fisher Scoring iterations: 2\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glm.sv3 <- glm(pData(inpData_sv)[,\"sv3\"] ~ pData(inpData_sv)[,\"LibraryPool\"] + \n",
    "               pData(inpData_sv)[,\"TIN\"] + pData(inpData_sv)[,\"RIN\"] +\n",
    "               pData(inpData_sv)[,\"DuplicationRate\"] + pData(inpData_sv)[,\"Etiology\"] +\n",
    "               pData(inpData_sv)[,\"Race\"] + pData(inpData_sv)[,\"Sex\"] + pData(inpData_sv)[,\"Weight\"] +\n",
    "               pData(inpData_sv)[,\"Height\"] + pData(inpData_sv)[,\"HW\"] + pData(inpData_sv)[,\"AFib\"] +\n",
    "               pData(inpData_sv)[,\"VTVF\"] + pData(inpData_sv)[,\"Diabetes\"] + pData(inpData_sv)[,\"Hypertension\"] +\n",
    "               pData(inpData_sv)[,\"LVEF\"]) \n",
    "summary(glm.sv3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = pData(inpData_sv)[, \"sv4\"] ~ pData(inpData_sv)[, \n",
       "    \"LibraryPool\"] + pData(inpData_sv)[, \"TIN\"] + pData(inpData_sv)[, \n",
       "    \"RIN\"] + pData(inpData_sv)[, \"DuplicationRate\"] + pData(inpData_sv)[, \n",
       "    \"Etiology\"] + pData(inpData_sv)[, \"Race\"] + pData(inpData_sv)[, \n",
       "    \"Sex\"] + pData(inpData_sv)[, \"Weight\"] + pData(inpData_sv)[, \n",
       "    \"Height\"] + pData(inpData_sv)[, \"HW\"] + pData(inpData_sv)[, \n",
       "    \"AFib\"] + pData(inpData_sv)[, \"VTVF\"] + pData(inpData_sv)[, \n",
       "    \"Diabetes\"] + pData(inpData_sv)[, \"Hypertension\"] + pData(inpData_sv)[, \n",
       "    \"LVEF\"])\n",
       "\n",
       "Deviance Residuals: \n",
       "      Min         1Q     Median         3Q        Max  \n",
       "-0.124318  -0.024523  -0.000557   0.021854   0.143812  \n",
       "\n",
       "Coefficients: (1 not defined because of singularities)\n",
       "                                                      Estimate Std. Error\n",
       "(Intercept)                                          0.0345176  0.0230448\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02          0.0770188  0.0153328\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03         -0.0424131  0.0140922\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04         -0.0069284  0.0142300\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05         -0.0141081  0.0148421\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06         -0.0484767  0.0159886\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07         -0.0437140  0.0141532\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08         -0.0238010  0.0135434\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09         -0.0477815  0.0150138\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10         -0.0220640  0.0146543\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11         -0.0430187  0.0146894\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12         -0.0552889  0.0143657\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                0.0122397  0.0097830\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                0.0119896  0.0110541\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                0.0269405  0.0112556\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                 -0.0181196  0.0081723\n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                 -0.0178397  0.0080799\n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                  -0.0258800  0.0095214\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421] -0.0008775  0.0085348\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517] -0.0176817  0.0104506\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935] -0.0046338  0.0129201\n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                  -0.0148139  0.0119293\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                  -0.0214009  0.0134729\n",
       "pData(inpData_sv)[, \"Race\"]AA                        0.0086982  0.0064095\n",
       "pData(inpData_sv)[, \"Sex\"]F                          0.0042469  0.0096174\n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]              -0.0051311  0.0078605\n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                 0.0057986  0.0086518\n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                0.0049899  0.0093630\n",
       "pData(inpData_sv)[, \"Height\"](163,170]              -0.0160202  0.0084422\n",
       "pData(inpData_sv)[, \"Height\"](170,178]              -0.0135502  0.0109565\n",
       "pData(inpData_sv)[, \"Height\"](178,196]              -0.0118346  0.0117438\n",
       "pData(inpData_sv)[, \"HW\"](372,448]                   0.0099415  0.0084178\n",
       "pData(inpData_sv)[, \"HW\"](448,558]                   0.0155617  0.0087488\n",
       "pData(inpData_sv)[, \"HW\"](558,923]                   0.0152792  0.0091540\n",
       "pData(inpData_sv)[, \"AFib\"]Yes                      -0.0055708  0.0068277\n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                       0.0094056  0.0070395\n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                  -0.0173719  0.0070108\n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                      NA         NA\n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                0.0059538  0.0088966\n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                0.0081862  0.0096405\n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                0.0055024  0.0128616\n",
       "                                                    t value Pr(>|t|)    \n",
       "(Intercept)                                           1.498 0.135593    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02           5.023 1.04e-06 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03          -3.010 0.002918 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04          -0.487 0.626821    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05          -0.951 0.342868    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06          -3.032 0.002719 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07          -3.089 0.002267 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08          -1.757 0.080230 .  \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09          -3.183 0.001669 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10          -1.506 0.133582    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11          -2.929 0.003760 ** \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12          -3.849 0.000155 ***\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                 1.251 0.212206    \n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                 1.085 0.279260    \n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                 2.394 0.017519 *  \n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                  -2.217 0.027623 *  \n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                  -2.208 0.028273 *  \n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                   -2.718 0.007084 ** \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421]  -0.103 0.918203    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517]  -1.692 0.092062 .  \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935]  -0.359 0.720197    \n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                   -1.242 0.215617    \n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                   -1.588 0.113610    \n",
       "pData(inpData_sv)[, \"Race\"]AA                         1.357 0.176131    \n",
       "pData(inpData_sv)[, \"Sex\"]F                           0.442 0.659219    \n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]               -0.653 0.514583    \n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                  0.670 0.503415    \n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                 0.533 0.594611    \n",
       "pData(inpData_sv)[, \"Height\"](163,170]               -1.898 0.059044 .  \n",
       "pData(inpData_sv)[, \"Height\"](170,178]               -1.237 0.217495    \n",
       "pData(inpData_sv)[, \"Height\"](178,196]               -1.008 0.314679    \n",
       "pData(inpData_sv)[, \"HW\"](372,448]                    1.181 0.238862    \n",
       "pData(inpData_sv)[, \"HW\"](448,558]                    1.779 0.076652 .  \n",
       "pData(inpData_sv)[, \"HW\"](558,923]                    1.669 0.096501 .  \n",
       "pData(inpData_sv)[, \"AFib\"]Yes                       -0.816 0.415418    \n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                        1.336 0.182880    \n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                   -2.478 0.013962 *  \n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                   NA       NA    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                 0.669 0.504050    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                 0.849 0.396716    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                 0.428 0.669201    \n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for gaussian family taken to be 0.001781954)\n",
       "\n",
       "    Null deviance: 0.77750  on 261  degrees of freedom\n",
       "Residual deviance: 0.39559  on 222  degrees of freedom\n",
       "  (92 observations deleted due to missingness)\n",
       "AIC: -876.35\n",
       "\n",
       "Number of Fisher Scoring iterations: 2\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glm.sv4 <- glm(pData(inpData_sv)[,\"sv4\"] ~ pData(inpData_sv)[,\"LibraryPool\"] + \n",
    "               pData(inpData_sv)[,\"TIN\"] + pData(inpData_sv)[,\"RIN\"] +\n",
    "               pData(inpData_sv)[,\"DuplicationRate\"] + pData(inpData_sv)[,\"Etiology\"] +\n",
    "               pData(inpData_sv)[,\"Race\"] + pData(inpData_sv)[,\"Sex\"] + pData(inpData_sv)[,\"Weight\"] +\n",
    "               pData(inpData_sv)[,\"Height\"] + pData(inpData_sv)[,\"HW\"] + pData(inpData_sv)[,\"AFib\"] +\n",
    "               pData(inpData_sv)[,\"VTVF\"] + pData(inpData_sv)[,\"Diabetes\"] + pData(inpData_sv)[,\"Hypertension\"] +\n",
    "               pData(inpData_sv)[,\"LVEF\"]) \n",
    "summary(glm.sv4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "glm(formula = pData(inpData_sv)[, \"sv5\"] ~ pData(inpData_sv)[, \n",
       "    \"LibraryPool\"] + pData(inpData_sv)[, \"TIN\"] + pData(inpData_sv)[, \n",
       "    \"RIN\"] + pData(inpData_sv)[, \"DuplicationRate\"] + pData(inpData_sv)[, \n",
       "    \"Etiology\"] + pData(inpData_sv)[, \"Race\"] + pData(inpData_sv)[, \n",
       "    \"Sex\"] + pData(inpData_sv)[, \"Weight\"] + pData(inpData_sv)[, \n",
       "    \"Height\"] + pData(inpData_sv)[, \"HW\"] + pData(inpData_sv)[, \n",
       "    \"AFib\"] + pData(inpData_sv)[, \"VTVF\"] + pData(inpData_sv)[, \n",
       "    \"Diabetes\"] + pData(inpData_sv)[, \"Hypertension\"] + pData(inpData_sv)[, \n",
       "    \"LVEF\"])\n",
       "\n",
       "Deviance Residuals: \n",
       "      Min         1Q     Median         3Q        Max  \n",
       "-0.283466  -0.023588   0.002147   0.025983   0.140069  \n",
       "\n",
       "Coefficients: (1 not defined because of singularities)\n",
       "                                                      Estimate Std. Error\n",
       "(Intercept)                                          0.0110040  0.0274284\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02          0.0073946  0.0182494\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03         -0.0035256  0.0167729\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04         -0.0087480  0.0169369\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05         -0.0090834  0.0176654\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06         -0.0103942  0.0190300\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07         -0.0052256  0.0168454\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08         -0.0156289  0.0161196\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09         -0.0772024  0.0178697\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10         -0.0284588  0.0174418\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11          0.0140046  0.0174836\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12          0.0038068  0.0170984\n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]               -0.0034439  0.0116439\n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                0.0217765  0.0131568\n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                0.0239190  0.0133966\n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                 -0.0077522  0.0097268\n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                 -0.0147513  0.0096169\n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                  -0.0147546  0.0113325\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421] -0.0102984  0.0101583\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517] -0.0104641  0.0124385\n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935] -0.0128092  0.0153778\n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                   0.0163560  0.0141985\n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                   0.0341647  0.0160357\n",
       "pData(inpData_sv)[, \"Race\"]AA                        0.0011668  0.0076287\n",
       "pData(inpData_sv)[, \"Sex\"]F                         -0.0156325  0.0114468\n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]               0.0073708  0.0093557\n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                 0.0033091  0.0102976\n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                0.0129066  0.0111440\n",
       "pData(inpData_sv)[, \"Height\"](163,170]               0.0008332  0.0100481\n",
       "pData(inpData_sv)[, \"Height\"](170,178]              -0.0083189  0.0130407\n",
       "pData(inpData_sv)[, \"Height\"](178,196]              -0.0072504  0.0139777\n",
       "pData(inpData_sv)[, \"HW\"](372,448]                   0.0151088  0.0100191\n",
       "pData(inpData_sv)[, \"HW\"](448,558]                  -0.0063403  0.0104130\n",
       "pData(inpData_sv)[, \"HW\"](558,923]                   0.0083815  0.0108952\n",
       "pData(inpData_sv)[, \"AFib\"]Yes                      -0.0094609  0.0081264\n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                      -0.0087933  0.0083785\n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                   0.0100664  0.0083444\n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                      NA         NA\n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                0.0121519  0.0105889\n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]               -0.0147806  0.0114744\n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]               -0.0036427  0.0153081\n",
       "                                                    t value Pr(>|t|)    \n",
       "(Intercept)                                           0.401   0.6887    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_02           0.405   0.6857    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_03          -0.210   0.8337    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_04          -0.517   0.6060    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_05          -0.514   0.6076    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_06          -0.546   0.5855    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_07          -0.310   0.7567    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_08          -0.970   0.3333    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_09          -4.320 2.35e-05 ***\n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_10          -1.632   0.1042    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_11           0.801   0.4240    \n",
       "pData(inpData_sv)[, \"LibraryPool\"]Magnet_12           0.223   0.8240    \n",
       "pData(inpData_sv)[, \"TIN\"](60.6,70.1]                -0.296   0.7677    \n",
       "pData(inpData_sv)[, \"TIN\"](70.1,73.3]                 1.655   0.0993 .  \n",
       "pData(inpData_sv)[, \"TIN\"](73.3,81.5]                 1.785   0.0756 .  \n",
       "pData(inpData_sv)[, \"RIN\"](8.1,8.5]                  -0.797   0.4263    \n",
       "pData(inpData_sv)[, \"RIN\"](8.5,8.9]                  -1.534   0.1265    \n",
       "pData(inpData_sv)[, \"RIN\"](8.9,10]                   -1.302   0.1943    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.378,0.421]  -1.014   0.3118    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.421,0.517]  -0.841   0.4011    \n",
       "pData(inpData_sv)[, \"DuplicationRate\"](0.517,0.935]  -0.833   0.4058    \n",
       "pData(inpData_sv)[, \"Etiology\"]DCM                    1.152   0.2506    \n",
       "pData(inpData_sv)[, \"Etiology\"]HCM                    2.131   0.0342 *  \n",
       "pData(inpData_sv)[, \"Race\"]AA                         0.153   0.8786    \n",
       "pData(inpData_sv)[, \"Sex\"]F                          -1.366   0.1734    \n",
       "pData(inpData_sv)[, \"Weight\"](66.2,79]                0.788   0.4316    \n",
       "pData(inpData_sv)[, \"Weight\"](79,91]                  0.321   0.7482    \n",
       "pData(inpData_sv)[, \"Weight\"](91,267]                 1.158   0.2480    \n",
       "pData(inpData_sv)[, \"Height\"](163,170]                0.083   0.9340    \n",
       "pData(inpData_sv)[, \"Height\"](170,178]               -0.638   0.5242    \n",
       "pData(inpData_sv)[, \"Height\"](178,196]               -0.519   0.6045    \n",
       "pData(inpData_sv)[, \"HW\"](372,448]                    1.508   0.1330    \n",
       "pData(inpData_sv)[, \"HW\"](448,558]                   -0.609   0.5432    \n",
       "pData(inpData_sv)[, \"HW\"](558,923]                    0.769   0.4425    \n",
       "pData(inpData_sv)[, \"AFib\"]Yes                       -1.164   0.2456    \n",
       "pData(inpData_sv)[, \"VTVF\"]Yes                       -1.050   0.2951    \n",
       "pData(inpData_sv)[, \"Diabetes\"]Yes                    1.206   0.2290    \n",
       "pData(inpData_sv)[, \"Hypertension\"]Yes                   NA       NA    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.15,0.2]                 1.148   0.2524    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.2,0.52]                -1.288   0.1990    \n",
       "pData(inpData_sv)[, \"LVEF\"](0.52,0.8]                -0.238   0.8121    \n",
       "---\n",
       "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1\n",
       "\n",
       "(Dispersion parameter for gaussian family taken to be 0.002524363)\n",
       "\n",
       "    Null deviance: 0.75948  on 261  degrees of freedom\n",
       "Residual deviance: 0.56041  on 222  degrees of freedom\n",
       "  (92 observations deleted due to missingness)\n",
       "AIC: -785.1\n",
       "\n",
       "Number of Fisher Scoring iterations: 2\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "glm.sv5 <- glm(pData(inpData_sv)[,\"sv5\"] ~ pData(inpData_sv)[,\"LibraryPool\"] + \n",
    "               pData(inpData_sv)[,\"TIN\"] + pData(inpData_sv)[,\"RIN\"] +\n",
    "               pData(inpData_sv)[,\"DuplicationRate\"] + pData(inpData_sv)[,\"Etiology\"] +\n",
    "               pData(inpData_sv)[,\"Race\"] + pData(inpData_sv)[,\"Sex\"] + pData(inpData_sv)[,\"Weight\"] +\n",
    "               pData(inpData_sv)[,\"Height\"] + pData(inpData_sv)[,\"HW\"] + pData(inpData_sv)[,\"AFib\"] +\n",
    "               pData(inpData_sv)[,\"VTVF\"] + pData(inpData_sv)[,\"Diabetes\"] + pData(inpData_sv)[,\"Hypertension\"] +\n",
    "               pData(inpData_sv)[,\"LVEF\"]) \n",
    "summary(glm.sv5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Principal Variance Component Analysis of the raw data with the surrogate variables included as covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# discretize the continuous surrogate variables \n",
    "var_names <- c(\"sv1\", \"sv2\", \"sv3\", \"sv4\", \"sv5\") \n",
    "pData(inpData_sv) <- conTocat(pData(inpData_sv), var_names) \n",
    "## Include the surrogate variables as covariates in addition to others\n",
    "cvrts_eff_var <- c(cvrts_eff_var, \"sv1_cat\", \"sv2_cat\", \"sv3_cat\", \"sv4_cat\", \"sv5_cat\")\n",
    "## again set the PVCA Threshold to explain 75% of the expression variation\n",
    "pct_thrsh <- 0.75 \n",
    "## Perform PVCA\n",
    "pvcaObj_sva <- pvcaBatchAssess(inpData_sv, cvrts_eff_var, pct_thrsh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(file.path(dirloc, 'PVCA_sva.png'), width=800, height=400)\n",
    "bp <- barplot(pvcaObj_sva$dat,\n",
    "              ylab=expression(bold(\"Weighted Average Proportion Variance\")),\n",
    "              ylim=c(0,1.1), col=c(\"navy\"), las=2, font=4,\n",
    "              main=\"Principal Variant Component Analysis Estimation\")\n",
    "axis(1,at=bp,labels=pvcaObj_sva$label,cex.axis=1,las=2,font=4)\n",
    "values <- pvcaObj_sva$dat\n",
    "new_values <- round(values,3)\n",
    "text(bp,pvcaObj_sva$dat,labels=new_values,pos=3,cex=1,font=4)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n",
      "boundary (singular) fit: see ?isSingular\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# For comparison, remove TIN and LibraryPool to see how much of it the surrogate variables capture\n",
    "pvcaObj_sva2 <- pvcaBatchAssess(inpData_sv, cvrts_eff_var[-which(cvrts_eff_var %in% c('LibraryPool', 'TIN'))], pct_thrsh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(file.path(dirloc, 'PVCA_sva1-2.png'), width=800, height=400)\n",
    "bp <- barplot(pvcaObj_sva2$dat,\n",
    "              ylab=expression(bold(\"Weighted Average Proportion Variance\")),\n",
    "              ylim=c(0,1.1), col=c(\"navy\"), las=2, font=4,\n",
    "              main=\"Principal Variant Component Analysis Estimation\")\n",
    "axis(1,at=bp,labels=pvcaObj_sva2$label,cex.axis=1,las=2,font=4)\n",
    "values <- pvcaObj_sva2$dat\n",
    "new_values <- round(values,3)\n",
    "text(bp,pvcaObj_sva2$dat,labels=new_values,pos=3,cex=1,font=4)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In summary, SV1 and SV2 capture a lot of the variance that was in LibraryPool and TIN..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R4.0.5",
   "language": "R",
   "name": "ir4.0.5"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
